Fabric.Authorization.API/scripts/AzureAD-Migration.ps1

```diff
@@ -0,0 +1,252 @@
+    return $groups;
+}
+
+function Get-AzureADGroupBySID {
+    param(
+        [Parameter(Mandatory=$true)]
+        [string] $tenantId,
+        [Parameter(Mandatory=$true)]
+        [string] $groupSID
+    )
+
+    $azureADGroups = @()
+    # connect to Azure AD
+    if ($null -ne $credentials[$tenantId]) {
+        Write-DosMessage -Level Information "Using cached credentials to connect to tenant $($tenantId)"
+        Connect-AzureADTenant -tenantId $tenantId -credential $credentials[$tenantId]
+    } else {
+        Write-DosMessage -Level Information "Credentials not cached - prompting user for credentials to connect to tenant $($tenantId)"
+        $authenticationFailed = $true
+        do {
+            try {
+                $credential = Connect-AzureADTenant -tenantId $tenantId
+                if (!$credentials.ContainsKey($tenantId)) {
+                    Write-DosMessage -Level Information -Message "Caching credentials for tenant $($tenantId)."
+                    $credentials.Add($tenantId, $credential)
+                }
+
+                Write-DosMessage -Level "Information" -Message "Retrieving group $($groupSID) from Azure AD Tenant $($tenantId)..."
+                $azureADGroups = Get-AzureADGroup -Filter "onPremisesSecurityIdentifier eq '$($groupSID)'"
+                $authenticationFailed = $false
+            }
+            catch [Microsoft.Open.Azure.AD.CommonLibrary.AadAuthenticationFailedException], [Microsoft.Open.Azure.AD.CommonLibrary.AadNeedAuthenticationException] {
+                Write-DosMessage -Level Error -Message "Error authenticating user with Azure AD tenant $($tenantId). Please try again."
+                $authenticationFailed = $true
+            }
+            # Connect-AzureADTenant will not throw an exception if a valid username (i.e., user@domain, even if user is not in tenant
+            # and/or invalid password is entered). This catch block handles the 403 that is returned when Get-AzureADGroup fails
+            # for this scenario
+            catch [Microsoft.Open.AzureAD16.Client.ApiException] {
+                if ($_.Exception.ErrorCode -eq 403) {
+                    Write-DosMessage -Level Information -Message "403 error when retrieving Azure AD group."
+                    # reset cached credentials for this tenantId
+                    $credentials.Remove($tenantId)
+                    Write-DosMessage -Level "Information" -Message "Removed cached credentials for Azure AD Tenant $($tenantId)..."
+                    $authenticationFailed = $true
+                }
+                else {
+                    Write-DosMessage -Level Information -Message "Unexpected error when retrieving Azure AD group. $($_.Exception)"
+                    throw
+                }
+            }
+            catch {
+                $errorMsg = "Unexpected error while connecting to Azure AD: $($_.Exception)"
+                Write-DosMessage -Level Error -Message $errorMsg
```

Comment:
If we want to throw the error, we can simply use Write-DosMessage -Level Fatal -Message $errorMsq